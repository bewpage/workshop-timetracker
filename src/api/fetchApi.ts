import { ClientRequestArgs } from 'http';
import { OperationTaskType, TaskType } from '../components/App/App';

export const options: ClientRequestArgs = {
  method: 'GET',
  headers: {
    Authorization: process.env.REACT_APP_API_KEY,
    'Content-Type': 'application/json',
  },
};

const baseAPI = async (url: string, options: any): Promise<any> => {
  const response = await fetch(process.env.REACT_APP_API_HOST + url, options);
  return await response.json();
};

export const getTasksAPI = async (
  options: any,
  url: string = '/api/tasks'
): Promise<any> => {
  // const url = '/api/tasks';
  return await baseAPI(url, options);
};

const getOperationAPI = async (
  options: any,
  id: string,
  url: string = `/api/tasks/${id}/operations`
): Promise<any> => {
  // const url = `/api/tasks/${id}/operations`;
  return await baseAPI(url, options);
};

const createTaskAPI = async (payload: TaskType): Promise<any> => {
  const option = {
    ...options,
    method: 'POST',
    body: JSON.stringify(payload),
  };
  return await getTasksAPI(option);
};

const updateTaskAPI = async (id: string, payload: TaskType): Promise<any> => {
  const option = {
    ...options,
    method: 'PUT',
    body: JSON.stringify(payload),
  };
  return await getTasksAPI(option, `/api/tasks/${id}`);
};

const deleteTaskAPI = async (id: string): Promise<any> => {
  const option = {
    ...options,
    method: 'DELETE',
  };
  return await getTasksAPI(option, `/api/tasks/${id}`);
};

const createOperationAPI = async (id: string, payload: any): Promise<any> => {
  const option = {
    ...options,
    method: 'POST',
    body: JSON.stringify(payload),
  };
  return await getOperationAPI(option, id);
};

const deleteOperationAPI = async (id: string): Promise<any> => {
  const option = {
    ...options,
    method: 'DELETE',
  };
  return await getOperationAPI(option, id, `/api/operations/${id}`);
};

const updateOperationAPI = async (
  id: string,
  payload: OperationTaskType
): Promise<any> => {
  const option = {
    ...options,
    method: 'PUT',
    body: JSON.stringify(payload),
  };
  return await getOperationAPI(option, id, `/api/operations/${id}`);
};

export const api = {
  getTasks: () => getTasksAPI(options),
  createTask: (payload: TaskType) => createTaskAPI(payload),
  updateTask: (id: string, payload: TaskType) => updateTaskAPI(id, payload),
  deleteTask: (id: string) => deleteTaskAPI(id),
  getOperations: (id: string) => getOperationAPI(options, id),
  createOperation: (id: string, payload: OperationTaskType) =>
    createOperationAPI(id, payload),
  updateOperation: (id: string, payload: OperationTaskType) =>
    updateOperationAPI(id, payload),
  deleteOperation: (id: string) => deleteOperationAPI(id),
};
