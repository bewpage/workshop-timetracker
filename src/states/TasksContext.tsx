import React, {
  createContext,
  PropsWithChildren,
  ReactNode,
  useEffect,
  useReducer,
} from 'react';
import { TaskType } from '../components/App/App';
import { ReducerEnumActions, taskReducer } from './taskReducer';
import { api } from '../api/fetchApi';

type Props = {
  children: ReactNode;
};

export type DataState = {
  isLoading: boolean;
  tasks: TaskType[] | [];
};

type TaskContextType = {
  data: DataState;
  dispatch: React.Dispatch<any>;
};

const INITIAL_STATE: DataState = {
  isLoading: false,
  tasks: [],
};

export const TaskContext = createContext<TaskContextType>({
  data: {
    isLoading: false,
    tasks: [],
  },
  dispatch: () => null,
});

const TasksContext: React.FC<PropsWithChildren<Props>> = ({ children }) => {
  const [data, dispatch] = useReducer(taskReducer, INITIAL_STATE);

  useEffect(() => {
    dispatch({ type: ReducerEnumActions.isLoading, payload: true });
    api
      .getTasks()
      .then(data => {
        if (!data.error) {
          dispatch({ type: ReducerEnumActions.getTasks, payload: data });
          return data.data;
        }
      })
      .then(data => {
        data.forEach((task: TaskType) => {
          api.getOperations(task.id).then(response => {
            if (!response.error) {
              dispatch({
                type: ReducerEnumActions.getOperations,
                payload: {
                  ...task,
                  operations: response.data,
                },
              });
            }
          });
        });
      })
      .catch(error => {
        // TODO handle error
        throw new Error(error);
      })
      .finally(() => {
        dispatch({ type: ReducerEnumActions.isLoading, payload: false });
      });
  }, []);

  return (
    <TaskContext.Provider value={{ data, dispatch }}>
      {children}
    </TaskContext.Provider>
  );
};

export default TasksContext;
