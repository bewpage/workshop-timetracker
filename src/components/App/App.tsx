import React from 'react';
import NewTaskForm from '../NewTaskForm/NewTaskForm';
import TaskList from '../TaskList/TaskList';
import TasksContext from '../../states/TasksContext';

export type OperationTaskType = {
  id: string;
  description: string;
  addedDate: string;
  timeSpent: number;
  task: TaskType;
};

export type TaskType = {
  id: string;
  title: string;
  description: string;
  status: string;
  addedDate: string;
  operations: OperationTaskType[] | [];
};

function App(): React.JSX.Element {
  return (
    <main id="app" className="js-tasks-list container mb-5">
      <TasksContext>
        <NewTaskForm />
        <TaskList />
      </TasksContext>
    </main>
  );
}

export default App;
