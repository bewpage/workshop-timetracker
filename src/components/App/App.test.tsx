import { render, screen } from '@testing-library/react';
import App from './App';
import { api } from '../../api/fetchApi';
import {
  mockAPIDataTest,
  operationMockData,
} from '../../states/mocks/taskMockData';
import TasksContext from '../../states/TasksContext';

describe('App', () => {
  beforeEach(() => {
    jest.spyOn(api, 'getTasks').mockResolvedValue({ data: [mockAPIDataTest] });
    jest
      .spyOn(api, 'getOperations')
      .mockResolvedValue({ data: [operationMockData] });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders App component', async () => {
    const { container } = render(
      <TasksContext>
        <App />
      </TasksContext>
    );

    await screen.findByText('Task Title Test');

    expect(screen.getByText(/New task/i)).toBeInTheDocument();
    expect(screen.getByText(/Task Title Test/i)).toBeInTheDocument();
    expect(screen.getByText(/Task Description Test/i)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
