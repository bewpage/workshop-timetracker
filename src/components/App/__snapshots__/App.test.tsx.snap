// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component 1`] = `
<div>
  <main
    class="js-tasks-list container mb-5"
    id="app"
  >
    <div
      class="card shadow"
    >
      <div
        class="card-body"
      >
        <h1
          class="card-title"
        >
          New task
        </h1>
        <form
          class="js-task-adding-form"
        >
          <div
            class="form-group"
          >
            <label
              class="sr-only"
              for="title"
            />
            <input
              class="form-control"
              id="title"
              minlength="5"
              name="title"
              placeholder="Title"
              required=""
              type="text"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="sr-only"
              for="description"
            />
            <input
              class="form-control"
              id="description"
              minlength="5"
              name="description"
              placeholder="Description"
              required=""
              type="text"
            />
          </div>
          <button
            class="btn btn-info"
            type="submit"
          >
            Add task
            <i
              class="fas fa-plus-circle ml-1"
            />
          </button>
        </form>
      </div>
    </div>
    <section
      class="card mt-5 shadow-sm"
    >
      <div
        class="card-header d-flex justify-content-between align-items-center"
      >
        <div>
          <h5>
            Task Title Test
          </h5>
          <h6
            class="card-subtitle text-muted"
          >
            Task Description Test
          </h6>
        </div>
        <div>
          <button
            class="btn btn-dark btn-sm"
          >
            Finish
          </button>
          <button
            class="btn btn-outline-danger btn-sm ml-2"
          >
            Delete
          </button>
        </div>
      </div>
      <ul
        class="list-group list-group-flush"
      >
        <li
          class="list-group-item d-flex justify-content-between align-items-center"
        >
          <div
            id="task-operation-item-description"
          >
            Description Operation Task Test
            <span
              class="badge badge-success badge-pill ml-2"
            >
               20m
            </span>
          </div>
          <div>
            <button
              class="btn btn-outline-success btn-sm mr-2"
            >
              +15m
            </button>
            <button
              class="btn btn-outline-success btn-sm mr-2"
            >
              +1h
            </button>
            <button
              class="btn btn-outline-danger btn-sm"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
      <div
        class="card-body"
      >
        <form>
          <div
            class="input-group"
          >
            <label
              class="sr-only"
              for="operationDescription"
            />
            <input
              class="form-control"
              id="operationDescription"
              minlength="5"
              name="description"
              placeholder="Operation description"
              required=""
              type="text"
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-info"
                type="submit"
              >
                Add
              </button>
            </div>
          </div>
        </form>
      </div>
    </section>
  </main>
</div>
`;
